File: Data Structures & Algorithms/validate-parentheses/submission-4.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;
        unordered_map<char, char> mp ={
            {'}', '{'},
            {']', '['},
            {')', '('},
        };
        for(auto c:s){
            if(mp.count(c)){
                if(!stack.empty() && stack.top()==mp[c]){
                stack.pop();
                }else{
                    return false;
                }
            }
            else{
                stack.push(c);
            }



        }
        return stack.empty();
    }
};

File: Data Structures & Algorithms/minimum-stack/submission-4.py
class MinStack:

    def __init__(self):
        self.minstack = []
        self.stack = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if self.minstack and val > self.minstack[-1]:
            self.minstack.append(self.minstack[-1])
            return
        self.minstack.append(val)


    def pop(self) -> None:
        self.stack.pop()
        self.minstack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.minstack[-1]
        